File: java/v2/src/main/java/com/kmeans/CSVReader.java
package com.kmeans;

import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;

public class CSVReader {

    private static final int SCALAR_FOR_WORKERS_READ = 1;
    private static final int ONE_MB = 1048576;
    private record Segment(long start, long size) {}

    private static List<Segment> getSegments(Path path) throws IOException {
        List<Segment> segments = new ArrayList<>();
        try (RandomAccessFile raf = new RandomAccessFile(path.toFile(), "r")) {
            long totalSize = raf.length();
            int cores = Runtime.getRuntime().availableProcessors() * SCALAR_FOR_WORKERS_READ;
            long targetSize = Math.max(totalSize / cores, ONE_MB);
            long currentPos = 0;

            while (currentPos < totalSize) {
                long start = currentPos;
                long endCandidate = currentPos + targetSize;

                if (endCandidate >= totalSize) {
                    segments.add(new Segment(start, totalSize - start));
                    break;
                }

                raf.seek(endCandidate);
                while (raf.getFilePointer() < totalSize && raf.read() != '\n');
                
                long actualEnd = raf.getFilePointer();
                segments.add(new Segment(start, actualEnd - start));
                currentPos = actualEnd;
            }
        }
        return segments;
    }

    public static List<Point> loadPoints(Path path) throws Exception {
        List<Segment> segments = getSegments(path);
        int numSegments = segments.size();
        @SuppressWarnings("unchecked")
        List<Point>[] partialResults = new List[numSegments];
        Thread[] threads = new Thread[numSegments];

        for (int i = 0; i < numSegments; i++) {
            final int index = i;
            Segment segment = segments.get(i);

            threads[i] = Thread.ofPlatform().start(() -> {
                List<Point> pointsPartition = new ArrayList<Point>();
                try (FileChannel ch = FileChannel.open(path, StandardOpenOption.READ)) {
                    ByteBuffer buffer = ch.map(FileChannel.MapMode.READ_ONLY, segment.start(), segment.size());

                    if (segment.start() == 0) {
                        while (buffer.hasRemaining() && buffer.get() != '\n');
                    }

                    StringBuilder lineBuilder = new StringBuilder();
                    while (buffer.hasRemaining()) {
                        byte b = buffer.get();
                        if (b == '\n') {
                            processLine(lineBuilder.toString(), pointsPartition);
                            lineBuilder.setLength(0);
                        } else if (b != '\r') {
                            lineBuilder.append((char) b);
                        }
                    }
                    if (lineBuilder.length() > 0) {
                        processLine(lineBuilder.toString(), pointsPartition);
                    }
                    partialResults[index] = pointsPartition;
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        }

        List<Point> points = new ArrayList<>();
        for (int i = 0; i < threads.length; i++) {
            threads[i].join();
            if (partialResults[i] != null) {
                points.addAll(partialResults[i]);
            }
        }

        return points;
    }

    private static void processLine(String currentLine, List<Point> points) {
        if (currentLine.isEmpty()) return;
        try {
            String[] parts = currentLine.split(",");
            double[] coords = new double[parts.length];
            for (int i = 0; i < parts.length; i++) {
                coords[i] = Double.parseDouble(parts[i].trim());
            }
            points.add(new Point(coords));
        } catch (NumberFormatException e) {
        }
    }
}

File: go/v3/csv_reader.go
package main

import (
	"bytes"
	"os"
	"runtime"
	"strconv"
	"sync"
)

type Segment struct {
	start int64
	size  int64
}

func Segments(filePath string) ([]Segment, error) {
	const oneMB = 1048576
	const scalarForWorkers = 1

	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	info, err := file.Stat()
	if err != nil {
		return nil, err
	}
	totalSize := info.Size() 
	cores := runtime.NumCPU() * scalarForWorkers

	targetSize := totalSize / int64(cores)
	if targetSize < oneMB {
		targetSize = oneMB
	}

	var segments []Segment
	var currentPos int64 = 0

	for currentPos < totalSize {
		start := currentPos
		endCandidate := currentPos + targetSize

		if endCandidate >= totalSize {
			segments = append(segments, Segment{start: start, size: totalSize - start})
			break
		}

		_, err := file.Seek(endCandidate, 0)
		if err != nil {
			return nil, err
		}

		buffer := make([]byte, 1)
		for {
			_, err := file.Read(buffer)
			if err != nil {
				break
			}
			if buffer[0] == '\n' {
				break
			}
		}

		actualEnd, _ := file.Seek(0, 1)
		segments = append(segments, Segment{start: start, size: actualEnd - start})
		currentPos = actualEnd
	}

	return segments, nil
}

func LoadPoints(path string) ([]Point, error) {
    segments, err := Segments(path)
    if err != nil {
        return nil, err
    }
    numSegments := len(segments)
    partialResults := make ([][]Point, numSegments)
    var wg sync.WaitGroup

    file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

    for i := 0; i < numSegments; i++ {
        wg.Add(1)
        go func (index int, segment Segment)  {
            defer wg.Done()
            
            var pointsPartition []Point
			buffer := make([]byte, segment.size)

			_, err := file.ReadAt(buffer, segment.start)
			if err != nil {
				return
			}

			remaining := buffer
			if segment.start == 0 {
				if idx := bytes.IndexByte(remaining, '\n'); idx != -1 {
					remaining = remaining[idx+1:]
				}
			}

			lines := bytes.Split(remaining, []byte("\n"))
			for _, line := range lines {
				line = bytes.TrimSpace(line)
				if len(line) > 0 {
					p, err := processLine(line)
                    if err != nil {
                        return
                    }
					if p != nil {
						pointsPartition = append(pointsPartition, *p)
					}
				}
			}
			partialResults[index] = pointsPartition
            
        }(i, segments[i])
    }

    wg.Wait()

    totalPoints := 0
    for _, partition := range partialResults {
        totalPoints += len(partition)
    }

    points := make([]Point, 0, totalPoints)
    for _, partition := range partialResults {
        if partition != nil {
            points = append(points, partition...)
        }
    }

    return points, nil
}

func processLine(line []byte) (*Point, error) {
	parts := bytes.Split(line, []byte(","))
	coords := make([]float64, len(parts))
	
	for i, part := range parts {
		val, err := strconv.ParseFloat(string(bytes.TrimSpace(part)), 64)
		if err != nil {
			return nil, err
		}
		coords[i] = val
	}

    point, err := NewPoint(coords)
    if err != nil {
        return nil, err
    }

	return point, nil
}
